pkg/operator/ceph/file/mirror/controller.go

```diff
@@ -0,0 +1,263 @@
+	err = c.Watch(&source.Kind{Type: &cephv1.CephFilesystemMirror{TypeMeta: controllerTypeMeta}}, &handler.EnqueueRequestForObject{}, opcontroller.WatchControllerPredicate())
+	if err != nil {
+		return err
+	}
+
+	// Watch all other resources
+	for _, t := range objectsToWatch {
+		err = c.Watch(&source.Kind{Type: t}, &handler.EnqueueRequestForOwner{
+			IsController: true,
+			OwnerType:    &cephv1.CephFilesystemMirror{},
+		}, opcontroller.WatchPredicateForNonCRDObject(&cephv1.CephFilesystemMirror{TypeMeta: controllerTypeMeta}, mgr.GetScheme()))
+		if err != nil {
+			return err
+		}
+	}
+
+	// Build Handler function to return the list of ceph object
+	// This is used by the watchers below
+	handlerFunc, err := opcontroller.ObjectToCRMapper(mgr.GetClient(), &cephv1.CephFilesystemMirrorList{}, mgr.GetScheme())
+	if err != nil {
+		return err
+	}
+
+	// Watch for CephCluster Spec changes that we want to propagate to us
+	err = c.Watch(&source.Kind{Type: &cephv1.CephCluster{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       opcontroller.ClusterResource.Kind,
+			APIVersion: opcontroller.ClusterResource.APIVersion,
+		},
+	},
+	}, handler.EnqueueRequestsFromMapFunc(handlerFunc), opcontroller.WatchCephClusterPredicate())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile reads that state of the cluster for a CephFilesystemMirror object and makes changes based on the state read
+// and what is in the CephFilesystemMirror.Spec
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileFilesystemMirror) Reconcile(context context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// workaround because the rook logging mechanism is not compatible with the controller-runtime logging interface
+	reconcileResponse, err := r.reconcile(request)
+	if err != nil {
+		updateStatus(r.client, request.NamespacedName, k8sutil.FailedStatus)
+		logger.Errorf("failed to reconcile %v", err)
+	}
+
+	return reconcileResponse, err
+}
+
+func (r *ReconcileFilesystemMirror) reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Fetch the CephFilesystemMirror instance
+	filesystemMirror := &cephv1.CephFilesystemMirror{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, filesystemMirror)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.Debug("CephFilesystemMirror resource not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, errors.Wrap(err, "failed to get CephFilesystemMirror")
+	}
+
+	// The CR was just created, initializing status fields
+	if filesystemMirror.Status == nil {
+		updateStatus(r.client, request.NamespacedName, k8sutil.Created)
+	}
+
+	// Make sure a CephCluster is present otherwise do nothing
```

Comment:
The mirror daemon doesn't require an mds to be running? In that case there is just nothing to replicate?

Reply:
I'm not sure to understand, we do need a CephCluster otherwise we cannot create a cephx key and authenticate, etc.

Reply:
Let's say they create the CephFilesystemMirroring CR, but don't create any CephFilesystem CRs. The mirror daemon will be able to start, correct? So there is no dependency on a filesystem.

Reply:
That's correct. It can run without any MDS around. I guess I was confused with the portion of the code you commented on.